niworkflows/interfaces/utils.py

```diff
@@ -256,3 +257,150 @@ def _gen_reference(fixed_image, moving_image, fov_mask=None, out_file=None,
+        except ValueError:
+            pass
+
+        # Matching affines
+        matching_affines = valid_qform and np.allclose(img.get_qform(), img.get_sform())
+
+        save_file = False
+
+        # Both match, qform valid (implicit with match), codes okay -> do nothing, empty report
+        if matching_affines and qform_code > 0 and sform_code > 0:
+            self._results['out_file'] = self.inputs.in_file
+            open(out_report, 'w').close()
+
+        # Row 2:
+        elif valid_qform and qform_code > 0:
+            img.set_sform(img.get_qform(), qform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: sform matrix set'
+            description = """\
+<p class="elem-desc">The sform has been copied from qform.</p>
+"""
+        # Rows 3-4:
+        # Note: if qform is not valid, matching_affines is False
+        elif sform_code > 0 and (not matching_affines or qform_code == 0):
+            img.set_qform(img.get_sform(), sform_code)
+            save_file = True
+            warning_txt = 'Note on orientation: qform matrix overwritten'
+            description = """\
+<p class="elem-desc">The qform has been copied from sform.</p>
+"""
+            if not valid_qform and qform_code > 0:
+                warning_txt = 'WARNING - Invalid qform information'
+                description = """\
+<p class="elem-desc">
+    The qform matrix found in the file header is invalid.
+    The qform has been copied from sform.
+    Checking the original qform information from the data produced
+    by the scanner is advised.
+</p>
+"""
+        # Rows 5-6:
+        else:
+            affine = img.affine
+            img.set_sform(affine, nb.nifti1.xform_codes['scanner'])
+            img.set_qform(affine, nb.nifti1.xform_codes['scanner'])
+            save_file = True
+            warning_txt = 'WARNING - Missing orientation information'
+            description = """\
+<p class="elem-desc">
+    Orientation information could not be retrieved from the image header.
+    The qform and sform matrices have been set to a default, LAS-oriented affine.
+    Analyses of this dataset MAY BE INVALID.
+</p>
+"""
+
+        if (self.inputs.force_float32 and np.dtype(img.get_data_dtype()).itemsize > 4) or self.inputs.n_volumes_to_discard:
+            in_data = img.get_data()
+
+            if (self.inputs.force_float32 and np.dtype(img.get_data_dtype()).itemsize > 4):
+                in_data = in_data.astype(np.float32)
+
+            img = nb.Nifti1Image(in_data[:, :, :, self.inputs.n_volumes_to_discard:],
+                                     img.affine,
+                                     img.header)
+            save_file = True
+
+        # Store new file and report
+        if save_file:
+            snippet = '<h3 class="elem-title">%s</h3>\n%s\n' % (
+            warning_txt, description)
+            with open(out_report, 'w') as fobj:
+                fobj.write(indent(snippet, '\t' * 3))
```

Comment:
If you don't have a warning_txt and description (i.e. the file was fine, but you wanted to coerce to float32), this will fail.

Reply:
Your interpretation is correct - as described in the input description. I left it this way because this is the way things have been done in the sanitizer function in MRIQC this interface is intended to replace.

Reply:
nice catch!